wave_lang/kernel/wave/codegen/read_write.py

```diff
@@ -747,179 +738,260 @@ def handle_tensor_load_to_lds(emitter: WaveEmitter, node: fx.Node):
+        tensor_dim_1_upper = rshift(dim_size_1, 16)
+        tile_size_0_shift = lshift(tile_size_0, 16)
+        pack = arith_d.ori(tensor_dim_1_upper, tile_size_0_shift)
+        d1 = vector_d.insert(pack, d1, static_position=[3], dynamic_position=[])
+
+        # tile size 1 is in good form
+        d1 = vector_d.insert(tile_size_1, d1, static_position=[4], dynamic_position=[])
+
+        # truncate upper 16 bit from dim stride 0 -> i48 to i32
+        dim_stride_0_trunc = arith_d.trunci(i32, dim_stride_0)
+        d1 = vector_d.insert(
+            dim_stride_0_trunc, d1, static_position=[5], dynamic_position=[]
+        )
+
+        # get upper 16 bit from dim stride 0, get lower 16 bit from dim stride 1, packed to i32
+        dim_stride_0_upper = rshift(dim_stride_0, 32)
+        dim_stride_0_trunc = arith_d.trunci(i32, dim_stride_0_upper)
+        dim_stride_1_lower = arith_d.trunci(i32, dim_stride_1)
+        dim_stride_1_trunc = lshift(dim_stride_1_lower, 16)
+        pack = arith_d.ori(dim_stride_0_trunc, dim_stride_1_trunc)
+        d1 = vector_d.insert(pack, d1, static_position=[6], dynamic_position=[])
+
+        # shift dim stride 1 to get upper 32 bit and pack to i32
+        dim_stride_1_sh = rshift(dim_stride_1, 16)
+        pack = arith_d.trunci(i32, dim_stride_1_sh)
+        d1 = vector_d.insert(pack, d1, static_position=[7], dynamic_position=[])
+
+        d0_results.append(d0)
+        d1_results.append(d1)
+        d2_results.append(d2)
+        d3_results.append(d3)
+
+    # Select the appropriate descriptors based on input_selector
+    # Build chained select operations for each descriptor
+    def select_descriptor(results_list, input_selector_val):
+        """Select from list of results using chained arith_d.select operations."""
+        assert len(results_list) > 0, "results_list must not be empty"
+        if len(results_list) == 1:
+            return results_list[0]
+
+        # Start with the last element as default
+        selected = results_list[-1]
+
+        # Chain selects from second-to-last backwards to first
+        for i in range(len(results_list) - 2, -1, -1):
+            # Create condition: selector_val == i
+            i_const = arith_d.constant(input_selector_val.type, i)
+            cond = arith_d.cmpi(arith_d.CmpIPredicate.eq, input_selector_val, i_const)
+            selected = arith_d.select(cond, results_list[i], selected)
+
+        return selected
+
+    input_selector_val = gen_sympy_index(subs, input_selector)
+    d0_selected = select_descriptor(d0_results, input_selector_val)
+    d1_selected = select_descriptor(d1_results, input_selector_val)
+    d2_selected = select_descriptor(d2_results, input_selector_val)
+    d3_selected = select_descriptor(d3_results, input_selector_val)
 
     # cpol
     cpol = arith_d.constant(i32, 0)
 
-    return llvm_d.call_intrinsic(
-        None, "llvm.amdgcn.tensor.load.to.lds", [d0, d1, d2, d3, cpol], [], []
+    llvm_d.call_intrinsic(
```

Comment:
Not necessarily in this PR but just as FYI we have https://mlir.llvm.org/docs/Dialects/ROCDLDialect/#rocdltensorloadtolds-rocdltensorloadtoldsop upstreamed. :)